DSWx-NI triggering logic

ryanghunter opened the pull request:
## Purpose

This PR adds DSWx-NI triggering logic support to the daac_data_subscriber system. It supports time based queries and native-id queries natively through the daac_data_subscriber top-level script.

## Proposed Changes
- [ADD] support for querying, cataloging, and triggering DSWx-NI jobs based on L2 NISAR GCOV products
- [ADD] `--query-replacement-file` options to daac_data_subscriber for using pre-canned json instead of query for testing

## Issues
- #872 

## Testing
- end-to-end manually tested triggering from mozart on dev cluster using UAT query results for L2 GCOV
   - validated that the expected 6 DSWx-NI jobs are triggered
- added unit tests for MGRS database logic and interactions

ryanghunter committed:
add mgrs track collection db obj and tests

ryanghunter committed:
update mgrs track queries, add tests for gcov query

ryanghunter committed:
update gcov catalog for simplicity

ryanghunter committed:
add smaller test case to gcov query

ryanghunter committed:
fix paths in gcov query test

ryanghunter committed:
update gcov w/ proper catalog query and submission

ryanghunter committed:
Merge branch 'develop' into dswxni-triggering-logic

ryanghunter committed:
refactor query logic to better match query pattern

ryanghunter committed:
add json query replacement option to data subscriber

ryanghunter committed:
add submit dswx, fix syntax issues

ryanghunter committed:
change queue in dswx-ni triggering logic

ryanghunter committed:
remove erroneous commit

ryanghunter committed:
use correct product path in job submission metadata

ryanghunter commented:
for MGRS db staging - 
add to `opera-ancillaries` s3 bucket under `dswx_ni/` directory, use `get_ancillary_from_s3` and use `s3 path` in settings.yaml and have fall back location filesystem

for testing lambda -
trigger via event-bridge timer for dswx-ni

ryanghunter committed:
fix metadata for job submission

ryanghunter committed:
load mgrs db from s3

ryanghunter committed:
fix db read, use args.job_queue

ryanghunter committed:
add lambda, event, and queue for dswx_ni query jobs

ryanghunter committed:
Merge branch 'develop' into dswxni-triggering-logic

ryanghunter committed:
add query job spec, hysdsio, and queues

ryanghunter committed:
fix formatting with hysds-io

chrisjrd reviewed (approved):
changes look good.
you may want to move the _extract* methods outside of the CmrQuery class (or just leave them in. i would organize them separately since they are low level, to keep the methods at a higher level of abstraction than parsing string/granule).
Some of the docstrings are redundant, and simply repeat the function name, so you could omit those docstrings without a loss in readability.

ryanghunter committed:
use correct queue for job

ryanghunter committed:
refactor _extract methods and update tests

ryanghunter committed:
use download job for dswxni

ryanghunter committed:
fix bugs

ryanghunter committed:
Merge branch 'develop' into dswxni-triggering-logic

ryanghunter committed:
correct permissions and queue

ryanghunter committed:
use correct provider for gcov download

ryanghunter committed:
use correct provider for gcov download

ryanghunter committed:
add asf gcov import

ryanghunter committed:
bug fix

ryanghunter committed:
missing imports

ryanghunter committed:
mgrs set and cycle id

ryanghunter committed:
fix import bug

ryanghunter committed:
use - instead of _ in batchids

ryanghunter committed:
add dswx_ni trigger from s3 products to end of download job

hhlee445 commented:
@ryanghunter getting this error

```
Traceback (most recent call last):
  File "/home/ops/verdi/ops/opera-pcm/util/exec_util.py", line 35, in wrapper
    status = func(*args, **kwargs)
  File "/home/ops/verdi/ops/opera-pcm/data_subscriber/daac_data_subscriber.py", line 45, in main
    run(sys.argv)
  File "/home/ops/verdi/ops/opera-pcm/data_subscriber/daac_data_subscriber.py", line 78, in run
    results["download"] = run_download(args, token, es_conn, netloc, username, password, cmr, job_id)
  File "/home/ops/verdi/ops/opera-pcm/data_subscriber/daac_data_subscriber.py", line 134, in run_download
    downloader.run_download(args, token, es_conn, netloc, username, password, cmr, job_id)
  File "/home/ops/verdi/ops/opera-pcm/data_subscriber/asf_gcov_download.py", line 31, in run_download
    return self.submit_dswx_ni_job_submission_handler(sets_to_process)
  File "/home/ops/verdi/ops/opera-pcm/data_subscriber/asf_gcov_download.py", line 35, in submit_dswx_ni_job_submission_handler
    jobs = self.trigger_dswx_ni_jobs(sets_to_process)
  File "/home/ops/verdi/ops/opera-pcm/data_subscriber/asf_gcov_download.py", line 84, in trigger_dswx_ni_jobs
    return [submit_dswx_ni_job(
  File "/home/ops/verdi/ops/opera-pcm/data_subscriber/asf_gcov_download.py", line 86, in <listcomp>
    job_queue=self.args.job_queue,
AttributeError: 'AsfDaacGcovDownload' object has no attribute 'args'
```

hhlee445 commented:
The following command should trigger three DSWx-NI jobs, since the track number 155 and frame number 67 are used by **_MS_155_64, MS_155_65, and MS_155_66_**:

```
python3 ~/mozart/ops/opera-pcm/data_subscriber/daac_data_subscriber.py query \
  -c NISAR_L2_GCOV_BETA_V1 \
  --job-queue=opera-job_worker-gcov_query \
  --chunk-size 1 \
  --native-id=NISAR_L2_PR_GCOV_016_156_A_011_2005_DVDV_A_20230701T000816_20230701T000835_T00408_N_P_J_001 \
  --endpoint UAT
```

chrisjrd committed:
cherry-pick

chrisjrd committed:
fix(DSWx-NI): query job only submitting a single download job rather than multiple

chrisjrd committed:
fix(DSWx-NI): query job only submitting a single download job rather than multiple

chrisjrd committed:
Merge branch 'refs/heads/develop' into dswxni-triggering-logic

chrisjrd committed:
fix(DSWx-NI): add mgrs_set_id to ID.

chrisjrd committed:
fix(DSWx-NI): UAT token parsing.

Using more flexible datetime parsing rather than introduce endpoint/env param to distinguish between UAT and OPS.

chrisjrd committed:
fix(DSWx-NI): reading args for params

chrisjrd committed:
fix(DSWx-NI): download job execution

chrisjrd committed:
Merge branch 'refs/heads/develop' into dswxni-triggering-logic

hhlee445 commented:
https://github.com/nasa/opera-sds-pcm/pull/1331
